newdle/client/src/Timeline.js

```diff
@@ -0,0 +1,145 @@
+import React from 'react';
+import Participant from './Participant';
+import PropTypes from 'prop-types';
+import _ from 'lodash';
+import {Header} from 'semantic-ui-react';
+import styles from './Timeline.module.scss';
+import moment from 'moment';
+
+function calculateWidth(start, end, minHour, maxHour) {
+  let startMins = start.hours() * 60 + start.minutes();
+  let endMins = end.hours() * 60 + end.minutes();
+
+  if (startMins < minHour * 60) {
+    startMins = minHour * 60;
+  }
+  if (endMins > maxHour * 60) {
+    endMins = maxHour * 60;
+  }
+
+  return ((endMins - startMins) / ((maxHour - minHour) * 60)) * 100;
+}
+
+function calculatePosition(start, minHour, maxHour) {
+  const spanMins = (maxHour - minHour) * 60;
+  let startMins = start.hours() * 60 + start.minutes() - minHour * 60;
+
+  if (startMins < 0) {
+    startMins = 0;
+  } else if (startMins >= spanMins) {
+    startMins = spanMins - 5;
+  }
+
+  return (startMins / spanMins) * 100;
+}
+
+function calculateBusyPositions(availability, minHour, maxHour) {
+  return availability.map(({participant, busySlots}) => {
+    const slots = busySlots.map(slot => {
+      const start = moment(slot.startTime, 'HH:mm');
+      const end = moment(slot.endTime, 'HH:mm');
+      const segmentWidth = calculateWidth(start, end, minHour, maxHour);
+      const segmentPosition = calculatePosition(start, minHour, maxHour);
+      return {
+        ...slot,
+        width: segmentWidth,
+        pos: segmentPosition,
+      };
+    });
+    return {
+      participant,
+      busySlots: slots,
+    };
+  });
+}
+
+function renderParticipant(participant) {
+  return (
+    <span className={styles['timeline-row-label']}>
+      {<Participant name={`${participant.firstName} ${participant.lastName}`} />}
+    </span>
+  );
+}
+
+function renderBusySlot({width, pos}, key) {
+  return (
+    <div
+      className={styles['busy-item']}
+      key={`timeline-busy-${key}`}
+      style={{left: `${pos}%`, width: `${width}%`}}
+    />
+  );
+}
+
+function renderBusyColumn({width, pos}) {
+  return <div className={styles['busy-column']} style={{left: `${pos}%`, width: `${width}%`}} />;
```

Comment:
There's a missing `key=` here:

I tried this out and it seems to work:
```js
function renderBusyColumn({ width, pos, startTime, endTime }) {
  return <div
    key={`${startTime}-${endTime}`}
    className={styles['busy-column']}
    style={{ left: `${pos}%`, width: `${width}%` }}
  />;
}
```

Reply:
Yes, thanks for pointing out. :) I took your idea and added participant id as well, since columns from different participants can have the same times.

i still get the warnings in the console though

Reply:
Solved